File: recommend/recommendClass.php
<?

class recommendClass
{

	private $redis;
	private $v_ns = '';
	private $j_ns = '';
	private $num_recommend_items;

	public function __construct()
	{

		$redis = new redis();
		$redis->connect('127.0.0.1', 6379);

		$this->redis = $redis;
		$this->num_recommend_items = 500;
		$this->v_ns = 'Viewer:Item:';
		$this->j_ns = 'Jaccard:Item:';
	}


	/**
	 * setRating
	 *
	 * @param int $user_id
	 * @param int $item_id
	 * @access public
	 * @return boolean
	 */
	public function setRating($user_id, $item_id)
	{
		$this->redis->lRem($this->v_ns . $item_id, $user_id);
		$this->redis->lPush($this->v_ns . $item_id, $user_id);
		$this->redis->lTrim($this->v_ns . $item_id, 0, $this->num_recommend_items - 1);
		return true;
	}

	/**
	 * calcJaccard
	 *
	 * @param int[] $all_item_ids
	 * @access public
	 * @return boolean
	 */
	public function calcJaccard($all_item_ids)
	{
		foreach ($all_item_ids as $item_id1) {
			$base = $this->redis->lRange($this->v_ns . $item_id1, 0, $this->num_recommend_items - 1);
			if (count($base) === 0) {
				continue;
			}
			foreach ($all_item_ids as $item_id2) {
				if ($item_id1 === $item_id2) {
					continue;
				}
				$target = $this->redis->lRange($this->v_ns . $item_id2, 0, $this->num_recommend_items - 1);
				if (count($target) === 0) {
					continue;
				}

				# calculation of jaccard
				$join = floatval(count(array_unique(array_merge($base, $target))));//unduplicated user_id
				$intersect = floatval(count(array_intersect($base, $target)));//Number of duplicate user_ids
				if (!$intersect || !$join) {
					continue;
				}
				$jaccard = $intersect / $join;

				$this->redis->zAdd($this->j_ns . $item_id1, $jaccard, $item_id2);//Set based on jaccard score
			}
		}
		return true;
	}

	/**
	 * getItems
	 *
	 * @param int $item_id
	 * @param int $num_recommended(optional)
	 * @access public
	 * @return int[]
	 */
	public function getItems($item_id, $num_recommended = -1)
	{
		//Get in descending order (highest to lowest similarity)
		return $this->redis->zRevRange($this->j_ns . $item_id, 0, $num_recommended);
	}
}

File: recommend/recommendClass.py
# -*- coding: utf-8 -*-
from __future__ import absolute_import
from __future__ import unicode_literals

class recommendClass:
    def __init__(self):
        import redis
        self.r = redis.Redis(host='localhost', port=6379, db=10)
    

    def calcJaccard(self,e1,e2):
        """
        :param e1: list of int
        :param e2: list of int
        :return: float
        """
        set_e1 = set(e1)
        set_e2 = set(e2)
        return float(len(set_e1 & set_e2)) / float(len(set_e1 | set_e2))
    

    def getKey(self,k):
        return 'JACCARD:PRODUCT:{}'.format(k)
    
    def setRating(self,products):
        """
        :pram products : str item_id : list of users_id
        """
        for key in products:
            base_customers = products[key]
            for key2 in products:
                if key == key2:
                    continue
                target_customers = products[key2]
                j = self.calcJaccard(base_customers,target_customers)
                self.r.zadd(self.getKey(key),{key2:j})

recommendClass = recommendClass();

#実際は各アイテムのセット数の上限を設ける
products = {
    'X': [1, 3, 5],
    'A': [2, 4, 5],
    'B': [1, 2, 3],
    'C': [2, 3, 4, 7],
    'D': [3],
    'E': [4, 6, 7],
}

recommendClass.setRating(products)
print recommendClass.r.zrevrange(recommendClass.getKey('A'),0,-1)
